app/services/qa/linked_data/deep_sort_service.rb

```diff
@@ -0,0 +1,244 @@
+            return -1 if !filtered.empty? && other_filtered.empty?
+            return 1 if filtered.empty? && !other_filtered.empty?
+            list_comparator(filtered, other_filtered)
+          end
+
+          # Walk through language sorted lists
+          # * for each language, determine how closely the list of terms matches
+          # * prioritize the list that gets the most low values
+          def multi_language_list_comparator(other)
+            combined_languages = languages.concat(other.languages).uniq
+            by_language_comparisons = {}
+            combined_languages.each do |lang|
+              cmp = list_comparator(filtered_literals(lang), other.filtered_literals(lang))
+              by_language_comparisons[lang] = cmp
+            end
+            cmp_sum = by_language_comparisons.values.sum
+            return 1 if cmp_sum.positive?
+            return -1 if cmp_sum.negative?
+            0
+          end
+
+          def list_comparator(list, other_list)
+            # if an element doesn't have any terms in a language, the other element sorts lower
+            return -1 if other_list.empty?
+            return 1 if list.empty?
+            shorter_list_size = [list.size, other_list.size].min
+            cmp = 0
+            0.upto(shorter_list_size - 1) do |idx|
+              cmp = to_downcase(list[idx]) <=> to_downcase(other_list[idx])
+              return cmp unless cmp.zero?
+            end
+            return cmp if list.size == other_list.size
+            other_list.size < list.size ? 1 : -1 # didn't find any diffs, shorter list is considered lower
+          end
+
+          def same_language?(lit, other_lit)
+            return false if only_one_has_language_marker?(lit, other_lit)
+            return true if neither_have_language_markers?(lit, other_lit)
+            lit.language == other_lit.language
+          end
+
+          def neither_have_language_markers?(lit, other_lit)
+            !language?(lit) && !language?(other_lit)
+          end
+
+          def only_one_has_language_marker?(lit, other_lit)
+            (!language?(lit) && language?(other_lit)) || (language?(lit) && !language?(other_lit))
+          end
+
+          def language?(lit)
+            language = lit.language if lit.respond_to?(:language)
+            language.present?
+          end
+
+          def preferred_language?(language)
+            preferred_language.present? ? language == preferred_language : false
+          end
+
+          def no_language?(language)
+            language.blank?
+          end
+
+          def to_downcase(lit)
+            lit.to_s.downcase
+          end
+
+          def filtered_literals_by_language
+            @filtered_literals_by_language ||= create_all_filters
+          end
+
+          def create_all_filters
+            bins = {}
+            0.upto(size - 1) do |idx|
+              lang = language(idx, literals)
+              filter = bins.key?(lang) ? bins[lang] : []
```

Comment:
This can also be simplified by fetch `filter = bins.fetch(lang, [])`

Reply:
Done